### PR TITLE
Feature/bb 52

### PR DESCRIPTION
**Pull request by YanaHV**

### CODE REVIEW

File: src/main/java/com/barterbay/app/controller/GoodController.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@AllArgsConstructor
+@RestController
+@RequestMapping("/good")
+public class GoodController {
+
+  private final GoodService goodService;
+
```
Comment (Paskar-Oleksandr) [inline src/main/java/com/barterbay/app/controller/GoodController.java:28]: please add one more method - **findAllGoods()**

File: src/main/java/com/barterbay/app/domain/Address.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import javax.persistence.OneToOne;
+import javax.persistence.SequenceGenerator;
+import javax.persistence.Table;
+import javax.validation.constraints.NotBlank;
+
+@Entity
+@Getter
+@Setter
+@Table(name = "address")
+public class Address extends AbstractAuditingEntity {
```
Comment (Paskar-Oleksandr) [inline src/main/java/com/barterbay/app/domain/Address.java]: No need `extends AbstractAuditingEntity`, we do not want audit Address table

File: src/main/java/com/barterbay/app/domain/Good.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    sequenceName = "good_sequence",
+    allocationSize = 1)
+  private Long id;
+
+  @NotBlank
+  @Column(name = "name", nullable = false, length = 100)
+  private String goodName;
+
+  @NotBlank
+  @Column(name = "description", nullable = false, unique = true, length = 200)
```
Comment (Paskar-Oleksandr) [inline src/main/java/com/barterbay/app/domain/Good.java]: `unique = true` - why description is unique?

Comment (Paskar-Oleksandr) [inline src/main/java/com/barterbay/app/domain/Good.java]: `@Column(name = "description"` - since your field has same name `description` - adding `name = "description"` is radundant

File: src/main/java/com/barterbay/app/domain/Good.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  private String description;
+
+  @NotBlank
+  @Column(name = "category", nullable = false)
```
Comment (Paskar-Oleksandr) [inline src/main/java/com/barterbay/app/domain/Good.java]: `name = "category"` - no need to specify name, because it will be default

File: src/main/java/com/barterbay/app/domain/Good.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  @Enumerated(EnumType.STRING)
+  private Category category;
+
+  @OneToOne(fetch = FetchType.LAZY)
+  @JoinColumn(name = "address_id", nullable = false)
+  private Address address;
+
+  @ManyToOne(fetch = FetchType.LAZY)
+  @JoinColumn(name = "user_id", nullable = false)
+  private User user;
```
Comment (Paskar-Oleksandr) [inline src/main/java/com/barterbay/app/domain/Good.java:63]: Since this entity used in User entity in a **Set** collection - you should implement **equals() and hashCode()**

image

Best practices for implementation **equals() and hashCode()** below - please, read all those articles

1. https://thorben-janssen.com/ultimate-guide-to-implementing-equals-and-hashcode-with-hibernate/
2. https://vladmihalcea.com/the-best-way-to-implement-equals-hashcode-and-tostring-with-jpa-and-hibernate/
3. https://vladmihalcea.com/hibernate-facts-equals-and-hashcode/

File: src/main/java/com/barterbay/app/domain/dto/good/GoodCreatedDTO.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+@Setter
+public class GoodCreatedDTO {
+  private Long id;
+  @NotBlank(message = "Good`s name cannot be empty")
+  private String goodName;
+  @NotBlank(message = "Description cannot be empty")
+  private String description;
+  @NotBlank(message = "Category cannot be empty")
+  private Category category;
+  private AddressDTO address;
```
Comment (Paskar-Oleksandr) [inline src/main/java/com/barterbay/app/domain/dto/good/GoodCreatedDTO.java]: address can not be null

File: src/main/java/com/barterbay/app/domain/dto/good/GoodCreatedDTO.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  private Long userOwnerId;
```
Comment (Paskar-Oleksandr) [inline src/main/java/com/barterbay/app/domain/dto/good/GoodCreatedDTO.java]: user id can not be null

File: src/main/java/com/barterbay/app/controller/GoodController.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  @GetMapping("/{goodId}")
+  public ResponseEntity<GoodCreatedDTO> getGoodById(@PathVariable long goodId) {
+    return ResponseEntity.ok(goodService.findGoodById(goodId));
+  }
+
+  @PutMapping("/{goodId}")
+  @ResponseStatus(HttpStatus.OK)
+  public void updateGoodById(@PathVariable long goodId,
+                             @RequestBody GoodCreatedDTO goodCreatedDTO) {
```
Comment (Paskar-Oleksandr) [inline src/main/java/com/barterbay/app/controller/GoodController.java]: ` @RequestBody GoodCreatedDTO goodCreatedDTO` - validation for this DTO will not work, because it missed annotation - `@Valid` - if you want to validate your DTO - you should add it

`@RequestBody @Valid GoodCreatedDTO goodCreatedDTO` - should be like this

File: src/main/java/com/barterbay/app/controller/GoodController.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  }
+
+  @DeleteMapping("/{goodId}")
+  @ResponseStatus(HttpStatus.NO_CONTENT)
+  public void deleteById(@PathVariable long goodId) {
+    goodService.deleteGoodById(goodId);
+  }
+
+  @PostMapping
+  public ResponseEntity<GoodCreatedDTO> createGood(@RequestBody GoodCreatedDTO goodCreatedDTO) {
```
Comment (Paskar-Oleksandr) [inline src/main/java/com/barterbay/app/controller/GoodController.java]: `@Valid`

File: src/main/java/com/barterbay/app/domain/Good.java (hunk 3 above)
Comment (Paskar-Oleksandr) [inline src/main/java/com/barterbay/app/domain/Good.java]: length = 200 - not enough
should be 1000 characters

File: src/main/java/com/barterbay/app/repository/GoodRepository.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+import org.springframework.data.jpa.repository.EntityGraph;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+import java.util.Optional;
+
+public interface GoodRepository extends JpaRepository<Good, Long> {
+
+  @Override
+  @EntityGraph(attributePaths = {"user"}, type = EntityGraph.EntityGraphType.LOAD)
+  Optional<Good> findById(Long aLong);
```
Comment (Paskar-Oleksandr) [inline src/main/java/com/barterbay/app/repository/GoodRepository.java]: Bad idea override general method `findById` and added EntityGraph there
If you need findById with EntityGraph - create a custom name, use HQL and add EntityGraph 
The method from JpaRepository - no need to override them, because they are widely used, and if you add eager loading for some of the - other team members probable will not about it, and they want Good without User, but they will get Good WITH user 

Why do you need eagerly fetch User for your good?
You GoodCreatedDTO does not contain any user DTO - why do you fetch user?)


image

File: src/main/java/com/barterbay/app/domain/dto/user/UserDTO.java [hunk 11]
```diff
@@ -1,23 +1,25 @@
...
-
-  @NotBlank(message = "First name can not be empty")
+  @NotBlank(message = "First name cannot be empty")
   private String firstName;
-
-  // TODO complete this DTO
+  private Long id;
+  private String lastName;
+  private String password;
+  private Set<GoodCreatedDTO> goods;
```
Comment (Paskar-Oleksandr) [inline src/main/java/com/barterbay/app/domain/dto/user/UserDTO.java]: `private Set<GoodCreatedDTO> goods;` - why did you add it here? Do you use this UserDTO somewhere and do you need goodDTO inside?

File: src/main/java/com/barterbay/app/servcie/GoodService.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+@AllArgsConstructor
+public class GoodService {
+
+  private final GoodRepository goodRepository;
+  private final GoodMapper goodMapper;
+  private final AddressMapper addressMapper;
+
+  @Transactional(readOnly = true)
+  public GoodCreatedDTO findGoodById(long goodId) {
+    return goodMapper.goodEntityToGoodDTO(goodRepository.findById(goodId).orElseThrow());
```
Comment (Paskar-Oleksandr) [inline src/main/java/com/barterbay/app/servcie/GoodService.java]: `orElseThrow()` - or else throw what? 
In case you can not find goods by id -  what customer will see? 100 line stack trace?
Should be thrown your custom exception, and that exception should be handled

Comment (Paskar-Oleksandr) [inline src/main/java/com/barterbay/app/servcie/GoodService.java]: `goodRepository.findById(goodId).orElseThrow()` - please, use it as a separate variable, like this ->

` 
    final var good = goodRepository.findById(goodId).orElseThrow();
    return goodMapper.goodEntityToGoodDTO(good);
  `

File: src/main/java/com/barterbay/app/servcie/GoodService.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  public void updateGoodById(long goodId, GoodCreatedDTO goodCreatedDTO) {
+    final Good goodById = goodRepository.findById(goodId).orElseThrow();
```
Comment (Paskar-Oleksandr) [inline src/main/java/com/barterbay/app/servcie/GoodService.java]: `final Good goodById` -> `final vargoodById`
same with exception as above

File: src/main/java/com/barterbay/app/servcie/GoodService.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    if (goodCreatedDTO.getGoodName() != null) {
```
Comment (Paskar-Oleksandr) [inline src/main/java/com/barterbay/app/servcie/GoodService.java]: all this if statements should be in the mapper, you are supposed to create a method in mapper which will do all checks and mappings
Your service SHOULD NOT do any mappings - for this we use MupStruct, move all mapping  logic to your mapper

File: src/main/resources/db/changelogs/004_add_user.xml [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+  xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+         http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd">
+
+  <changeSet id="004_add_user" author="Hvozdetska Yana">
+    <sql>
+      INSERT INTO bb_user
+      VALUES (1, 'u.email@gmail.com', 'Yana', 'Hvozdetska', 'UGFzc3dvcmRfcGFzc3dvcmQ=', 0, now(), now()),
+             (2, 'ffffffff.email@gmail.com', 'Tom', 'Smit', 'UGFzc3dvcmRfcGFzc3dvcmQx', 0, now(), now());
```
Comment (Paskar-Oleksandr) [inline src/main/resources/db/changelogs/004_add_user.xml]: why did you add those users? for what purpose?

File: src/main/resources/db/changelogs/sql/001_init_base_tables.sql [hunk 16]
```diff
@@ -17,3 +17,5 @@ CREATE TABLE bb_user
 
 CREATE UNIQUE INDEX ue_index ON bb_user (email);
 
+
```
Comment (Paskar-Oleksandr) [inline src/main/resources/db/changelogs/sql/001_init_base_tables.sql]: why this file contains 2 new lines?

File: src/main/resources/db/changelogs/sql/003_add_address_table.sql [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+  country            VARCHAR(50) NOT NULL,
+  city               VARCHAR(50) NOT NULL,
+  street             VARCHAR(50) NOT NULL,
+  zip_code           int         NOT NULL,
+  created_date       TIMESTAMP   NOT NULL,
+  last_modified_date TIMESTAMP   NOT NULL
+);
+
+INSERT INTO address
+VALUES (1, 'UKR', 'Kharkiv', 'Kalyna', 61000, now(), now());
```
Comment (Paskar-Oleksandr) [inline src/main/resources/db/changelogs/sql/003_add_address_table.sql]: why do you need to INSERT something here??

File: src/main/resources/db/changelogs/sql/005_add_goods_table.sql [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+  FOREIGN KEY (user_id) REFERENCES bb_user (user_id)
+);
+
+ALTER TABLE address
+  ADD COLUMN good BIGINT;
+ALTER TABLE address
+  ADD FOREIGN KEY (good) REFERENCES good (good_id);
+
+INSERT INTO good
+VALUES (1, 'Tomatoes', ' Fresh and testy', 'FOOD', 1, 1, now(), now());
```
Comment (Paskar-Oleksandr) [inline src/main/resources/db/changelogs/sql/005_add_goods_table.sql]: for what reason did you already create one good and added it to DB?

**Review by Paskar-Oleksandr — CHANGES_REQUESTED** [review]

File: src/main/java/com/barterbay/app/domain/Address.java (hunk 2 above)
Comment (YanaHV) [reply src/main/java/com/barterbay/app/domain/Address.java]: Corrected

File: src/main/java/com/barterbay/app/domain/Good.java (hunk 3 above)
Comment (YanaHV) [reply src/main/java/com/barterbay/app/domain/Good.java]: Corrected

Comment (YanaHV) [reply src/main/java/com/barterbay/app/domain/Good.java]: Corrected

File: src/main/java/com/barterbay/app/domain/Good.java (hunk 4 above)
Comment (YanaHV) [reply src/main/java/com/barterbay/app/domain/Good.java]: Corrected

File: src/main/java/com/barterbay/app/domain/dto/good/GoodCreatedDTO.java (hunk 6 above)
Comment (YanaHV) [reply src/main/java/com/barterbay/app/domain/dto/good/GoodCreatedDTO.java]: Corrected

File: src/main/java/com/barterbay/app/domain/dto/good/GoodCreatedDTO.java (hunk 7 above)
Comment (YanaHV) [reply src/main/java/com/barterbay/app/domain/dto/good/GoodCreatedDTO.java]: Corrected

**Review by Paskar-Oleksandr — APPROVED** [review]
